runtime/wasm/src/asc_abi/asc_ptr.rs

```diff
@@ -61,17 +67,16 @@ impl<C: AscType> AscPtr<C> {
     pub(crate) fn is_null(&self) -> bool {
         self.0 == 0
     }
-}
 
-impl<C> From<AscPtr<C>> for RuntimeValue {
-    fn from(ptr: AscPtr<C>) -> RuntimeValue {
-        RuntimeValue::from(ptr.0)
+    // Erase type information.
+    pub(crate) fn erase(self) -> AscPtr<()> {
+        AscPtr(self.0, PhantomData)
     }
 }
 
-impl<C> FromRuntimeValue for AscPtr<C> {
-    fn from_runtime_value(val: RuntimeValue) -> Option<Self> {
-        u32::from_runtime_value(val).map(|ptr| AscPtr(ptr, PhantomData))
+impl<C> From<i32> for AscPtr<C> {
+    fn from(ptr: i32) -> Self {
+        AscPtr(u32::from_le_bytes(ptr.to_le_bytes()), PhantomData)
```

Comment:
This is not the same because `ptr as u32` will panic in debug mode in case of overflow. This is used in the `link!` macro in a generic way so we need it to be a `From` impl.

I've added a comment explaining why we need this, though I see no fundamental reason of why wasmtime shouldn't allow `u32` as a primitive so I proposed this https://github.com/bytecodealliance/wasmtime/pull/1808/files upstream.

Reply:
Correction, `as` will never panic, as we discussed.

I switched this to using `try_from(ptr).unwrap()`.